### PR TITLE
feat: implemented FontWeightEditor

### PR DESCRIPTION
**Pull request by ernestchakhoyan**

### CODE REVIEW

File: src/internal/FontWeightEditor/FontWeightEditor.style.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+import { Property } from 'csstype';
+
+const useStyles = makeStyles<Theme>(({ zIndex }) =>
+    createStyles({
+        popper: {
+            zIndex: `${zIndex.tooltip + 11} !important` as Property.ZIndex, // more than the highest zIndex
+        },
+    })
+);
+
```
Comment (davityavryan) [inline src/internal/FontWeightEditor/FontWeightEditor.style.ts:11]: We need to make this as our theme part. :/
nothing for you to do. just a reminder

File: src/internal/FontWeightEditor/FontWeightEditor.tsx [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    { value: 'bold' },
+    { value: 'bolder' },
+    { value: 'lighter' },
+    { value: 'inherit' },
+    { value: 'initial' },
+    { value: 'unset' },
+];
+
+function FontWeightEditor({ value, onChange }: TProps) {
+    const [localValue, setLocalValue] = React.useState<TFontWeightValue | null>({ value: `${value}` as TFontWeight });
```
Comment (davityavryan) [inline src/internal/FontWeightEditor/FontWeightEditor.tsx:39]: 1. don't use `as` just use `{ value }` and extend `TProps` to be
```
type TProps = TextFieldProps & {
    value: TFontWeight;
};
```

2. import `useState`. as of some of the other named exports are imported that way.
3.  I can't still understand why you store `{ value }` structure. just store the value inside in `localValue` and that's it. This just complicated everything. and in that case you don't need `TFontWeightValue`

File: src/internal/FontWeightEditor/FontWeightEditor.tsx [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+            setLocalValue(fontWeight);
+        },
+        [onChange]
+    );
+
+    useEffect(() => {
+        // Update localValue if value is changed from outside(ex. reset)
+        if (timer === null && value !== localValue.value) {
+            setLocalValue({ value: `${value}` as TFontWeight });
```
Comment (davityavryan) [inline src/internal/FontWeightEditor/FontWeightEditor.tsx:109]: this `as` will not be needed anymore
```suggestion
            setLocalValue({ value });
```

File: src/internal/FontWeightEditor/FontWeightEditor.tsx [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    { value: 'unset' },
+];
+
+function FontWeightEditor({ value, onChange }: TProps) {
+    const [localValue, setLocalValue] = React.useState<TFontWeightValue | null>({ value: `${value}` as TFontWeight });
+
+    const [errorMessage, setErrorMessage] = React.useState<string>('');
+    const classes = useStyles();
+
+    const handleValidation = (value: string): string => {
```
Comment (davityavryan) [inline src/internal/FontWeightEditor/FontWeightEditor.tsx:44]: don't return `string`. Just show the error if needed. there will not be a need to have ternary anymore.
```
setErrorMessage(validationError);
```

File: src/internal/FontWeightEditor/FontWeightEditor.tsx [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+                    return;
+                }
+                const validationError = handleValidation(newValue.value);
+                if (validationError) {
+                    return setErrorMessage(validationError);
+                }
+                setValue(newValue);
+            }
+        },
+        [onChange]
```
Comment (davityavryan) [inline src/internal/FontWeightEditor/FontWeightEditor.tsx:88]: this depends on `handleValidation`

File: src/internal/FontWeightEditor/FontWeightEditor.tsx (hunk 4 above)
Comment (davityavryan) [inline src/internal/FontWeightEditor/FontWeightEditor.tsx:44]: useCallback

File: src/internal/FontWeightEditor/FontWeightEditor.tsx [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    { value: 'inherit' },
+    { value: 'initial' },
+    { value: 'unset' },
+];
+
+function FontWeightEditor({ value, onChange }: TProps) {
+    const [localValue, setLocalValue] = React.useState<TFontWeightValue | null>({ value: `${value}` as TFontWeight });
+
+    const [errorMessage, setErrorMessage] = React.useState<string>('');
+    const classes = useStyles();
```
Comment (davityavryan) [inline src/internal/FontWeightEditor/FontWeightEditor.tsx:42]: please move classes as first thing

File: src/internal/FontWeightEditor/FontWeightEditor.tsx [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    { value: '800' },
+    { value: '900' },
+    { value: 'normal' },
+    { value: 'bold' },
+    { value: 'bolder' },
+    { value: 'lighter' },
+    { value: 'inherit' },
+    { value: 'initial' },
+    { value: 'unset' },
+];
```
Comment (davityavryan) [inline src/internal/FontWeightEditor/FontWeightEditor.tsx:36]: let's move this inside with `useMemo`

File: src/internal/FontWeightEditor/FontWeightEditor.tsx [hunk 8]
```diff
@@ -0,0 +1,170 @@
+import React, { memo, useCallback, useEffect } from 'react';
+import { TextField, TextFieldProps } from '@material-ui/core';
+import Autocomplete, { createFilterOptions } from '@material-ui/lab/Autocomplete';
```
Comment (davityavryan) [inline src/internal/FontWeightEditor/FontWeightEditor.tsx:3]: add spaces each time when packages are not relevant.
import Autocomplete as named import

File: src/internal/FontWeightEditor/FontWeightEditor.tsx [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+            value={localValue}
+            onChange={handleChange}
+            filterOptions={(options, params) => {
+                const filtered = filter(options, params);
+
+                // Suggest the creation of a new value
+                if (params.inputValue !== '') {
+                    filtered.push({
+                        inputValue: params.inputValue,
+                        value: `Add "${params.inputValue}"` as TFontWeight,
```
Comment (davityavryan) [inline src/internal/FontWeightEditor/FontWeightEditor.tsx:131]: remove `as` as of it is incorrect. the problem why you need to do is in line 9. type should allow value as `TFontWeight | string`.

File: src/internal/FontWeightEditor/FontWeightEditor.tsx [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+                // Value selected with enter, right from the input
+                if (typeof option === 'string') {
+                    return option;
+                }
+                // Add "xxx" option created dynamically
+                if (option.inputValue) {
+                    return option.inputValue;
+                }
+                // Regular option
+                return option.value;
```
Comment (davityavryan) [inline src/internal/FontWeightEditor/FontWeightEditor.tsx:152]: not clear what is the difference between `Value selected with enter, right from the input` and `Regular option`

File: src/internal/FontWeightEditor/FontWeightEditor.tsx [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+            }}
+            renderOption={(option) => option.value}
+            style={{ width: 300 }}
```
Comment (davityavryan) [inline src/internal/FontWeightEditor/FontWeightEditor.tsx:155]: No no no inline style.

File: src/internal/FontWeightEditor/FontWeightEditor.tsx [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+            freeSolo
```
Comment (davityavryan) [inline src/internal/FontWeightEditor/FontWeightEditor.tsx:156]: boolean props should be at the end.

File: src/internal/FontWeightEditor/FontWeightEditor.tsx (hunk 11 above)
Comment (davityavryan) [reply src/internal/FontWeightEditor/FontWeightEditor.tsx:155]: never hardcode size in pixels.

File: src/utils/themeMap.tsx [hunk 13]
```diff
@@ -450,7 +450,7 @@ export const typesMap: TTypesMap = {
     ),
     easing: ({ value, onChange, ...props }) => <FieldEditor {...props} value={value} onChange={onChange} />,
     fontFamily: ({ value, onChange, ...props }) => <FieldEditor {...props} value={value} onChange={onChange} />,
-    fontWeight: ({ value, onChange, ...props }) => <FieldEditor {...props} value={value} onChange={onChange} />,
+    fontWeight: ({ value, onChange, ...props }) => <FontWeightEditor {...props} value={value} onChange={onChange} />,
```
Comment (davityavryan) [inline src/utils/themeMap.tsx:453]: can we simplify this to ???
```suggestion
    fontWeight: FontWeightEditor,
```

**Review by davityavryan — CHANGES_REQUESTED** [review]
you need to add in `babel.config.js`

```
    [
        'babel-plugin-import',
        {
            libraryName: '@material-ui/lab',
            libraryDirectory: isCommonJS ? '' : 'esm',
            camel2DashComponentName: false,
        },
        'lab',
    ],
```

So that tree shaking will work for lab.
See https://material-ui.com/guides/minimizing-bundle-size/

**Comment by davityavryan** [comment]
Also please do not push as just one commit. try to introduce gradual, logical commits.

File: src/internal/FontWeightEditor/FontWeightEditor.tsx [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+            renderInput={(params) => (
+                <TextField
+                    {...params}
+                    label="Font weight"
```
Comment (davityavryan) [inline src/internal/FontWeightEditor/FontWeightEditor.tsx:160]: use `label={themeKeyLabel(themeKey)}`

File: src/internal/FontWeightEditor/FontWeightEditor.tsx [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+                        value: `Add "${params.inputValue}"` as TFontWeight,
+                    });
+                }
+
+                return filtered;
+            }}
+            selectOnFocus
+            clearOnBlur
+            handleHomeEndKeys
+            id="fw-autocomplete"
```
Comment (davityavryan) [inline src/internal/FontWeightEditor/FontWeightEditor.tsx:140]: no need for this

File: src/internal/FontWeightEditor/FontWeightEditor.tsx [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+];
+
+function FontWeightEditor({ value, onChange }: TProps) {
+    const [localValue, setLocalValue] = React.useState<TFontWeightValue | null>({ value: `${value}` as TFontWeight });
+
+    const [errorMessage, setErrorMessage] = React.useState<string>('');
+    const classes = useStyles();
+
+    const handleValidation = (value: string): string => {
+        const valueIsNumber = +value;
```
Comment (davityavryan) [inline src/internal/FontWeightEditor/FontWeightEditor.tsx:45]: `Number.parseInt(value)`

File: src/internal/FontWeightEditor/FontWeightEditor.tsx (hunk 2 above)
Comment (haynar) [reply src/internal/FontWeightEditor/FontWeightEditor.tsx:39]: Totally agree I don't see the need of having a `{ value: '' }` structure. We can have an array of `string`s which will make the usage a breeze. In that case you won't need to repeat the logic in `handleChange` method for different types.

File: src/internal/FontWeightEditor/FontWeightEditor.tsx (hunk 9 above)
Comment (haynar) [reply src/internal/FontWeightEditor/FontWeightEditor.tsx:131]: I'd rather use `{ value: '', label: '' }` structure for clarity and just use map for options.

File: src/internal/FontWeightEditor/FontWeightEditor.tsx [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+            clearTimeout(timer);
+        };
+    }, []);
+
+    return (
+        <Autocomplete
+            classes={{ popper: classes.popper }}
+            value={localValue}
+            onChange={handleChange}
+            filterOptions={(options, params) => {
```
Comment (haynar) [inline src/internal/FontWeightEditor/FontWeightEditor.tsx:124]: Let's move the definition of this method either to `useCallback` or to outer scope so it won't create a new function on every render.

File: src/internal/FontWeightEditor/FontWeightEditor.tsx (hunk 10 above)
Comment (haynar) [reply src/internal/FontWeightEditor/FontWeightEditor.tsx:152]: This all cases seem very confusing, better to use the `{ value: '', label: '' }` structure as suggested above.

File: src/internal/FontWeightEditor/FontWeightEditor.tsx [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+                }
+
+                return filtered;
+            }}
+            selectOnFocus
+            clearOnBlur
+            handleHomeEndKeys
+            id="fw-autocomplete"
+            options={fontWeightValues}
+            getOptionLabel={(option) => {
```
Comment (haynar) [inline src/internal/FontWeightEditor/FontWeightEditor.tsx:142]: Let's move all the function props into a constant, as suggested above.

**Review by haynar — CHANGES_REQUESTED** [review]

File: src/internal/FontWeightEditor/FontWeightEditor.tsx (hunk 7 above)
Comment (davityavryan) [reply src/internal/FontWeightEditor/FontWeightEditor.tsx:36]: I don't have any arguments. It just feels lonely outside of component.

File: src/internal/FontWeightEditor/FontWeightEditor.tsx (hunk 9 above)
Comment (davityavryan) [reply src/internal/FontWeightEditor/FontWeightEditor.tsx:131]: This is not even needed. I was able to have same results with just value and inputValue.

File: src/internal/FontWeightEditor/FontWeightEditor.tsx (hunk 10 above)
Comment (davityavryan) [reply src/internal/FontWeightEditor/FontWeightEditor.tsx:152]: First condition can be removed by just not adding new suggestion as string and having it as inputValue(can be called) label probably. And so whole function becomes just return of that label or value

File: src/internal/FontWeightEditor/FontWeightEditor.tsx (hunk 7 above)
Comment (haynar) [reply src/internal/FontWeightEditor/FontWeightEditor.tsx:36]: IMO it's better having all the declarations that have no dependencies to outer scope to minimize the things that the component has to care about, so no `useCallback` or `useMemo` will be needed

**Pull request merged**